### DISCUSSION TITLE
Implementing stack unwinding

### DISCUSSION THREAD

**Discussion by tpdenk** [phil-opp/blog_os#1133] [Questions]
I'm trying to get stack unwinding implemented using https://github.com/nbdd0121/unwinding .
That requires a linker symbol `__etext` and `__eh_frame`, which are not provided by the rust linker. (flavour GNU LLD and `__GNU_EH_FRAME_HDR` doesn't work, since the linker still doesn't emit the symbol for some reason)
I'm having a hard time creating a link script that lets me still boot the kernel, since the boot loader fails to map pages (depends on the build script what mapping exactly fails), and I don't have any experience with writing link scripts.

The question is: How do I implement stack unwinding? Or at least what would a link script have to provide at which locations, to still work with the https://github.com/rust-osdev/bootloader , which I'm using?

If I'm running into the wrong direction, please let me know!

**Comment by bjorn3** [comment]
Just to confirm you have set `panic-strategy` to `unwind` in the target spec, right? Also try passing `--eh-frame-hdr` as linker argument. That should cause `__GNU_EH_FRAME_HDR` to be provided.

**Reply by tpdenk** [reply]
Correct. `--eh-frame-hdr` as linker argument does not provide a symbol `__GNU_EH_FRAME_HDR` on `ld64` (MacOS). That seems to be a known issue, but since I primarily develop on a Mac, I wouldn't like to have to compile in Docker.

The error I get is during compilation and reads like `linker symbol __etext / __eh_frame` not found.

**Reply by bjorn3** [reply]
You should be using ld.lld instead of ld64. ld.lld is an ELF linker. ld64 is a Mach-O linker. Bootloader only works with ELF binaries.

**Reply by tpdenk** [reply]
I'm lost again. I'm trying that right now.
### Cargo.toml
```toml
[dependencies.unwinding]
version = "0.1.4"
default-features = false
features = ["unwinder", "personality", "fde-gnu-eh-frame-hdr", "panic"]
```
### target JSON
```json
{
  "llvm-target": "x86_64-unknown-none",
  "data-layout": "e-m:e-i64:64-f80:128-n8:16:32:64-S128",
  "arch": "x86_64",
  "target-endian": "little",
  "target-pointer-width": "64",
  "target-c-int-width": "32",
  "os": "none",
  "executables": true,
  "linker-flavor": "ld.lld",
  "pre-link-args": {
    "ld.lld": [
      "--eh-frame-hdr"
    ]
  },
  "linker": "rust-lld",
  "panic-strategy": "unwind",
  "disable-redzone": true,
  "features": "-mmx,-sse,+soft-float"
}
```

### lib.rs
```rust
extern crate unwinding;
```

## `cargo build`
(After `cargo clean`)

```text
inlinable function call in a function with debug info must have a !dbg location
  call void @_Unwind_Resume(ptr %81) #14
inlinable function call in a function with debug info must have a !dbg location
  call void @_Unwind_Resume(ptr %101) #14
inlinable function call in a function with debug info must have a !dbg location
  call void @_Unwind_Resume(ptr %144) #14
inlinable function call in a function with debug info must have a !dbg location
  call void @_Unwind_Resume(ptr %111) #14
LLVM ERROR: Broken module found, compilation aborted!
libc++abi: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
libc++abi: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
libc++abi: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
error: could not compile `unwinding`

Caused by:
  process didn't exit successfully: `rustc --crate-name unwinding --edition=2018 /Users/tsatke/.cargo/registry/src/github.com-1ecc6299db9ec823/unwinding-0.1.4/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="alloc"' --cfg 'feature="fde-gnu-eh-frame-hdr"' --cfg 'feature="panic"' --cfg 'feature="panicking"' --cfg 'feature="personality"' --cfg 'feature="unwinder"' -C metadata=b28e6f54a98e6e59 -C extra-filename=-b28e6f54a98e6e59 --out-dir /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps --target /Users/tsatke/Development/github.com/martimos/kernel/x86_64-martim.json -L dependency=/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps -L dependency=/Users/tsatke/Development/github.com/martimos/kernel/target/debug/deps --extern 'noprelude:alloc=/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/liballoc-be0226ebf61ffacc.rmeta' --extern 'noprelude:compiler_builtins=/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcompiler_builtins-b8caa82bd2c12503.rmeta' --extern 'noprelude:core=/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcore-8ccff3e1ca1a321f.rmeta' --extern gimli=/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libgimli-c720d5ab9106cd4c.rmeta -Z unstable-options --cap-lints allow` (signal: 6, SIGABRT: process abort signal)
warning: build failed, waiting for other jobs to finish..
```

Did I forget something obvious? That seems like an LLVM error

**EDIT** Seems that with those features and the current nightly, even the unwinding crate itself can't be built, same error (`cargo build --no-default-features --features unwinder,fde-static,personality,panic`)

**Reply by bjorn3** [reply]
Seems to be a rustc or LLVM bug. As workaround I think you can disable debuginfo using the following in `Cargo.toml`:

```toml
[profile.dev]
debug = 0

[profile.release]
debug = 0
```

**Reply by tpdenk** [reply]
Now I got the old error again.

<details>
<summary>Full error</summary>
cargo b
   Compiling martim v0.1.0 (/Users/tsatke/Development/github.com/martimos/kernel)
error: linking with `rust-lld` failed: exit status: 1
  |
  = note: "rust-lld" "-flavor" "gnu" "--eh-frame-hdr" "/var/folders/8q/k0fjrmzd3pq24266vh_lpgrc0000gn/T/rustcdTo1c2/symbols.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.18f83gufxpnuhzwv.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.1m5fyokp83xbhd65.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.1s14ocg5eh88r6hl.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.1sw36b7oyoet1n9l.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.23jhhpsd6i2ykqip.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.243dcfnucz9eyscx.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.2gz03mrlircz77ub.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.2qmwowszfpozoiwd.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.2tp5p4vpp6hbpuus.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.2wuamfb6twymhnhi.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.2zi984d8lqkun8j0.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.3mkq4hf69cw338ir.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.3rwsry8nn07gf66i.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.3xgsjdlvwx1nblvu.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.415c4w1fgnyvnkgl.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.4iye53pc3ivg4xll.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.4nw7b46fyl7dmngf.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.4z0cpf7ow7oikx30.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.4zru2bd45xw0n5xl.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.53agq0sek6o78iv7.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.53n9hzk3q8iw40q0.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.540x858gs1narabz.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.5ai58t5alqjbcc6x.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.5ehb5fwioyk2aco8.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.5fyb9o56rs7mfnik.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.7ul1oosnmcu0we1.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.82553vfs9apmin.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.9oxhliym0abrq7s.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.oal8nt40qn6pqgl.rcgu.o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911.4boytl9dsmp4i97s.rcgu.o" "--as-needed" "-L" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps" "-L" "/Users/tsatke/Development/github.com/martimos/kernel/target/debug/deps" "-L" "/Users/tsatke/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/x86_64-martim/lib" "-Bstatic" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libmartim-ee61516665077449.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/liblinked_list_allocator-f20f0f81768eaece.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libspinning_top-f5b08bf39ede435a.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libnoto_sans_mono_bitmap-e7a25d9e72cda330.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libpc_keyboard-540f0e6b9fc6a5e5.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libfutures_util-f308cbeae853f055.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libpin_project_lite-91ba5756210f9591.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libfutures_task-37446cd9071990ce.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libpin_utils-e6786f5fa2647b63.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libfutures_core-e309bf4dc56200cf.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libconquer_once-a0c830253f705237.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcrossbeam_queue-616a2361f3c33e91.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcrossbeam_utils-d432a3d1d24bd3cd.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcfg_if-dc08127856ffaa4b.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libkernel_constants-56fc7b3b9951b7ce.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libuart_16550-762edde3044468aa.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libpic8259-39f4e3fcdf3fa4e7.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/liblazy_static-6183329ed1ae5709.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libspin-63f49f2ac6171b14.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libx86_64-3f1dc1f3221e201c.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libvolatile-fb45d84958a12832.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libbitflags-188c7fa18b85e6af.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libbit_field-1bcc97373793b5ff.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libkstd-6af96e4d0cdfd78a.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libspin-313c65d9ea76e027.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/liblock_api-326233d38bc6166e.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libscopeguard-af1bce1d6b066adc.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libgimli-7bea3e8cbb895daa.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libbootloader-ff8684e16fff74dc.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/liballoc-63e543bab37a4a12.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/librustc_std_workspace_core-98476eeb5bec2c9e.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcore-391cc2d89b35edca.rlib" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libcompiler_builtins-018094c474eb0434.rlib" "-Bdynamic" "--eh-frame-hdr" "-znoexecstack" "-L" "/Users/tsatke/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/x86_64-martim/lib" "-o" "/Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/martim-74deb89d94380911" "--gc-sections"
  = note: rust-lld: error: undefined symbol: __etext
          >>> referenced by unwinding.4edbead8-cgu.13
          >>>               unwinding-3fe4452918606348.unwinding.4edbead8-cgu.13.rcgu.o:(_$LT$unwinding..unwinder..find_fde..gnu_eh_frame_hdr..StaticFinder$u20$as$u20$unwinding..unwinder..find_fde..FDEFinder$GT$::find_fde::h31f8791ff0ece925) in archive /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib

          rust-lld: error: undefined symbol: __GNU_EH_FRAME_HDR
          >>> referenced by unwinding.4edbead8-cgu.13
          >>>               unwinding-3fe4452918606348.unwinding.4edbead8-cgu.13.rcgu.o:(_$LT$unwinding..unwinder..find_fde..gnu_eh_frame_hdr..StaticFinder$u20$as$u20$unwinding..unwinder..find_fde..FDEFinder$GT$::find_fde::h31f8791ff0ece925) in archive /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib
          >>> referenced by unwinding.4edbead8-cgu.13
          >>>               unwinding-3fe4452918606348.unwinding.4edbead8-cgu.13.rcgu.o:(_$LT$unwinding..unwinder..find_fde..gnu_eh_frame_hdr..StaticFinder$u20$as$u20$unwinding..unwinder..find_fde..FDEFinder$GT$::find_fde::h31f8791ff0ece925) in archive /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib


error: could not compile `martim` due to previous error
</details>

```
  = note: rust-lld: error: undefined symbol: __etext
          >>> referenced by unwinding.4edbead8-cgu.13
          >>>               unwinding-3fe4452918606348.unwinding.4edbead8-cgu.13.rcgu.o:(_$LT$unwinding..unwinder..find_fde..gnu_eh_frame_hdr..StaticFinder$u20$as$u20$unwinding..unwinder..find_fde..FDEFinder$GT$::find_fde::h31f8791ff0ece925) in archive /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib

          rust-lld: error: undefined symbol: __GNU_EH_FRAME_HDR
          >>> referenced by unwinding.4edbead8-cgu.13
          >>>               unwinding-3fe4452918606348.unwinding.4edbead8-cgu.13.rcgu.o:(_$LT$unwinding..unwinder..find_fde..gnu_eh_frame_hdr..StaticFinder$u20$as$u20$unwinding..unwinder..find_fde..FDEFinder$GT$::find_fde::h31f8791ff0ece925) in archive /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib
          >>> referenced by unwinding.4edbead8-cgu.13
          >>>               unwinding-3fe4452918606348.unwinding.4edbead8-cgu.13.rcgu.o:(_$LT$unwinding..unwinder..find_fde..gnu_eh_frame_hdr..StaticFinder$u20$as$u20$unwinding..unwinder..find_fde..FDEFinder$GT$::find_fde::h31f8791ff0ece925) in archive /Users/tsatke/Development/github.com/martimos/kernel/target/x86_64-martim/debug/deps/libunwinding-3fe4452918606348.rlib
```

In the full error you can see that `--eh-frame-hdr` is being passed in.

I think the problem is, that `rust-lld` uses another linker implementation.
```
rust-lld
lld is a generic driver.
Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead
```
And `ld64` doesn't care about `--eh-frame-hdr`.

**Reply by bjorn3** [reply]
> I think the problem is, that rust-lld uses another linker implementation.

The linker flavor is set to `ld.lld`, so it should act like an ELF linker. If you set the linker flavor to `ld64.lld` it would act like the `ld64` Mach-O linker.

> Now I got the old error again.

Maybe you need to declare those symbols in a linker script? See https://github.com/nbdd0121/unwinding#baremetal

**Reply by tpdenk** [reply]
Which brings us back to the original question, how such a linker script would look like. But I don't understand how that has to look like. I can create a simple one, and it lets me compile and even boot, but the bootloader immediately fails to map some pages and the kernel halts.

**Reply by bjorn3** [reply]
Unfortunately I don't know enough about linker scripts to help you.

**Reply by tpdenk** [reply]
Thank you so much anyways, especially for the quick replies. Really awesome that we could clear this up within a few hours!
